packages/typescript/src/core/entities/span.ts

```diff
@@ -0,0 +1,577 @@
+    // by the user. However, there is not way to set the status when using
+    // `trace` function wrapper. Therefore, we just automatically set the status
+    // to OK if it is not ERROR.
+    if (this.status.statusCode !== SpanStatusCode.ERROR) {
+      this.setStatus(SpanStatusCode.OK);
+    }
+
+    const endTime = options?.endTimeNs ? convertNanoSecondsToHrTime(options.endTimeNs) : undefined;
+    this._span.end(endTime);
+  }
+}
+
+/**
+ * A no-operation span implementation that doesn't record anything
+ */
+export class NoOpSpan implements LiveSpan {
+  readonly _span: any; // Use any for NoOp span to avoid type conflicts
+  readonly _attributesRegistry: _SpanAttributesRegistry;
+
+  constructor(span?: any) {
+    // Create a minimal no-op span object
+    this._span = span || {
+      spanContext: () => ({
+        spanId: '0000000000000000',
+        traceId: '00000000000000000000000000000000'
+      }),
+      attributes: {},
+      events: []
+    };
+    this._attributesRegistry = new _SpanAttributesRegistry(this._span as OTelSpan);
+  }
+
+  get traceId(): string {
+    return NO_OP_SPAN_TRACE_ID;
+  }
+  get spanId(): string {
+    return '';
+  }
+  get parentId(): string | null {
+    return null;
+  }
+  get name(): string {
+    return '';
+  }
+  get spanType(): SpanType {
+    return SpanType.UNKNOWN;
+  }
+  get startTime(): HrTime {
+    return [0, 0];
+  }
+  get endTime(): HrTime | null {
+    return null;
+  }
+  get status(): SpanStatus {
+    return new SpanStatus(SpanStatusCode.UNSET);
+  }
+  get inputs(): any {
+    return null;
+  }
+  get outputs(): any {
+    return null;
+  }
+  get attributes(): Record<string, any> {
+    return {};
+  }
+
+  getAttribute(_key: string): any {
+    return null;
+  }
+
+  // Implement all methods to do nothing
+  setInputs(_inputs: any): void {}
+  setOutputs(_outputs: any): void {}
+  setAttribute(_key: string, _value: any): void {}
+  setAttributes(_attributes: Record<string, any>): void {}
+  setStatus(_status: SpanStatus | SpanStatusCode | string, _description?: string): void {}
+  addEvent(_event: SpanEvent): void {}
+  recordException(_error: Error): void {}
+  end(
+    _outputs?: any,
+    _attributes?: Record<string, any>,
+    _status?: SpanStatus | SpanStatusCode,
+    _endTimeNs?: number
+  ): void {}
+
+  get events(): SpanEvent[] {
+    return [];
+  }
+
+  toJson(): any {
+    return {};
+  }
+}
+
+/**
+ * A utility class to manage the span attributes.
+ * In MLflow users can add arbitrary key-value pairs to the span attributes, however,
+ * OpenTelemetry only allows a limited set of types to be stored in the attribute values.
+ * Therefore, we serialize all values into JSON string before storing them in the span.
+ * This class provides simple getter and setter methods to interact with the span attributes
+ * without worrying about the serde process.
+ */
+class _SpanAttributesRegistry {
```

Comment:
In ts, I don't think we use _xxx naming convention since class/const cannot be imported unless it's exported.